feat(module): handle `tailwindMerge` config from `app.config`

stijns96 opened the pull request:
<!---
☝️ PR title should follow conventional commits (https://conventionalcommits.org)
-->

### 🔗 Linked issue

<!-- If it resolves an open issue, please link the issue here. For example "Resolves #123" -->

Resolves #1939 

### ❓ Type of change

<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->

- [x] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

<!-- Describe your changes in detail -->
<!-- Why is this change required? What problem does it solve? -->

This PR introduces the ability to override/extend `tailwind-merge` behaviour through the `app.config`. This is necessary when overriding the `tailwind.config` default values as explained here: https://github.com/dcastil/tailwind-merge/blob/v2.1.0/docs/configuration.md#usage-with-custom-tailwind-config

> More details on how to override the config: https://github.com/dcastil/tailwind-merge/blob/v2.1.0/docs/configuration.md#extending-the-tailwind-merge-config

When extending the spacing for example in your `tailwind.config.ts`:

```ts
import type { Config } from 'tailwindcss'

export default <Partial<Config>>{
  theme: {
    extend: {
      spacing: {
        sm: '0.5rem',
        base: '1rem',
        lg: '2rem'
      }
    }
  }
}

```

You can now configure tailwind-merge to merge those values with each other.

```ts
import { theme } from "#tailwind-config";

export default defineAppConfig({
  ui: {
    tailwindMerge: {
      override: {
        theme: {
          spacing: [...Object.keys(theme.spacing)],
        },
      },
    }
  },
})
```

### 📝 Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

stijns96 committed:
fix(module): handle `tailwindMerge` config from `app.config`

stijns96 commented:
Hi @benjamincanac,

I decided to ditch the restored branch and created `fix/extend-tailwind-merge` as the old one already contained 25 merge conflicts.

Additionally, I think it's almost fully working, but there is still one challenge. It looks like tailwind-merge doesn't export a type for the extendTailwindMerge function which I want to use for the `app.config.ts`. I opened a discussion on their repo.

I would like to to wait to merge this when we at least have an answer on this

stijns96 commented:
Hi @benjamincanac 

I found a "workaround". The only annoying thing is that it suggest to types... The normal one at least is working, but I can't figure out how to get rid of the second one. TW Merge haven't replied yet.

<img width="671" alt="image" src="https://github.com/user-attachments/assets/1ea97fea-35c3-4735-b0e9-7d0d723c3754" />

<img width="657" alt="image" src="https://github.com/user-attachments/assets/aa001261-47c6-4ebb-a17c-97d2837dfc0d" />


**./playgroud/src/module.ts**
```ts
type TailwindMergeConfig = Parameters<typeof extendTailwindMerge>[0];

type UI = {
  primary?: string
  gray?: string
  colors?: string[]
  strategy?: Strategy
  tailwindMerge?: TailwindMergeConfig
  [key: string]: any
} & DeepPartial<typeof config, string | number | boolean>
```

Any thoughts on this?

stijns96 committed:
fix(module): add `tailwindMerge` type to UI configuration

benjamincanac commented:
@stijns96 Seems fine! I just did the same work for `v3`: https://github.com/nuxt/ui/pull/2938

stijns96 commented:
@benjamincanac cool! Should I undraft the PR? When do you expect it to be released? I really need it in my MVP...

Let me know!

P.s. have to add some docs as well

benjamincanac commented on src/module.ts:
```diff
@@ -15,11 +16,15 @@ delete defaultColors.trueGray
 
+type TailwindMergeConfig = Parameters<typeof extendTailwindMerge>[0];
+
+
```
```suggestion

```

benjamincanac committed:
Update src/module.ts

benjamincanac commented:
I don't know when will the release happen but once it's merged on `dev`, you will be able to use the edge version anyway.

benjamincanac committed:
Merge branch 'dev' into fix/extend-tailwind-merge

benjamincanac commented:
Do you plan to make the docs in this PR?

benjamincanac committed:
lint

stijns96 commented:
@benjamincanac I think that might be the easiest to do. I'll try to do that asap (hopefully tonight).

benjamincanac commented:
I guess you can put it somewhere in https://ui.nuxt.com/getting-started/theming#components. I'll have to do it also on my PR for `v3`.

stijns96 commented:
> I guess you can put it somewhere in https://ui.nuxt.com/getting-started/theming#components. I'll have to do it also on my PR for `v3`.

Did you really mean under the "Components" or as a dedicated toc item? E.g. 

- Components
- Extend Tailwind Merge
- Dark mode

benjamincanac commented:
I would say under `Components` yes as we mention `tailwind-merge` inside `ui prop` and `class attribute`.

stijns96 committed:
docs(theming): add section on extending Tailwind Merge for custom configurations

This update introduces documentation on how to extend Tailwind Merge to accommodate custom Tailwind CSS configurations, including examples for `app.config.ts` and `tailwind.config.ts`. This enhancement aims to improve user understanding and implementation of Tailwind Merge in their projects.

stijns96 committed:
refactor(module): streamline TailwindMerge type definition and import

This commit refactors the module by removing the redundant import of `extendTailwindMerge` and consolidating the type definition for `TailwindMergeConfig`. This change enhances code clarity and maintains consistency in type usage.

stijns96 committed:
Merge branch 'fix/extend-tailwind-merge' of https://github.com/stijns96/ui into fix/extend-tailwind-merge

stijns96 committed:
refactor(module): remove redundant import of extendTailwindMerge

This commit simplifies the module by eliminating the duplicate import of `extendTailwindMerge`, enhancing code clarity and consistency.

stijns96 commented:
@benjamincanac There you go

<img width="788" alt="image" src="https://github.com/user-attachments/assets/6f2777f3-ade6-4c93-bb22-eb84140c5394" />

<img width="628" alt="image" src="https://github.com/user-attachments/assets/57a7de6c-8a24-4355-8d62-852a4ded61f3" />

stijns96 commented:
https://github.com/dcastil/tailwind-merge/pull/505

stijns96 committed:
Merge branch 'dev' into fix/extend-tailwind-merge

benjamincanac commented:
@stijns96 Do you want to wait for their release?

stijns96 commented:
@benjamincanac was thinking the same. Let me check with them when the are planning a release

benjamincanac committed:
Merge branch 'dev' into fix/extend-tailwind-merge

benjamincanac commented:
@stijns96 Should be good!

stijns96 committed:
chore(deps): update tailwind-merge to version 2.6.0 and adjust type definitions in module

This commit updates the `tailwind-merge` dependency to version 2.6.0 in both `package.json` and `pnpm-lock.yaml`. Additionally, it refines the type definitions in `src/module.ts` by replacing the import of `extendTailwindMerge` with a more specific type definition for `tailwindMerge`, enhancing type safety and clarity.

stijns96 committed:
Merge branch 'fix/extend-tailwind-merge' of https://github.com/stijns96/ui into fix/extend-tailwind-merge

stijns96 commented:
@benjamincanac,

Has been released and PR has been updated. Works like a 🍀 now!

benjamincanac commented:
Thanks @stijns96! 😊

stijns96 commented:
@benjamincanac you're welcome! When do you plan te release it? Looking forward to use it 😃
